### ISSUE TITLE
[CLIFF] About the gt 2d keypoint generation

### ISSUE THREAD

**Issue by Rainfalllen** [huawei-noah/noah-research#127]
Very elegant work! Could you share how to produce 2d keypoints grouth truth in full image?

**Comment by Rainfalllen** [comment]
> If you already got the 2d keypoint GT in the crop image, then just plus the displacement between the crop center and the image center.

Does 2d keypoint GT need to be normalized to [-1, 1] ?

**Comment by Rainfalllen** [comment]
> If you already got the 2d keypoint GT in the crop image, then just plus the displacement between the crop center and the image center.

Is the range of 2D keypoint GT in the cropped image in [-0.5*crop_size, 0.5*crop_size] ？

**Comment by Rainfalllen** [comment]
Is there something wrong with my code? I have exploding gradients at the beginning of training (2d loss is NaN).

```
    def j2d_processing_full(self, kp, center, scale, r, f, img_w, img_h):
        """For full image process gt 2D keypoints and apply all augmentation transforms."""
        nparts = kp.shape[0]
        for i in range(nparts):
            kp[i, 0:2] = transform(kp[i, 0:2] + 1, center, scale,
                                   [self.options.IMG_RES, self.options.IMG_RES], rot=r)

        # flip the x coordinates
        if f:
            kp = flip_kp(kp)

        img_center = [img_w / 2.0, img_h / 2.0]
        delta = center - img_center
        kp[:, :-1] = kp[:, :-1] + delta[None, :]
        kp[:, 0] = 2. * kp[:, 0] / img_w
        kp[:, 1] = 2. * kp[:, 1] / img_h

        kp = kp.astype('float32')
        return kp
```
```
        pred_cam_full = cam_crop2full(pred_cam_crop, bbox_center, bbox_scale,
                                          full_img_shape, focal_length)

        smpl_output = self.smpl(
            rotmat=hmr_output['pred_pose'],
            shape=hmr_output['pred_shape'],
        )

        joints3d = smpl_output['smpl_joints3d']
        bs = joints3d.shape[0]
        device = joints3d.device
        joints2d = perspective_projection(
            joints3d,
            rotation=torch.eye(3, device=device).unsqueeze(0).expand(bs, -1, -1),
            translation=pred_cam_full,
            focal_length=focal_length,
            camera_center=torch.zeros(bs, 2, device=device)
            # camera_center=img_center_full
        )

        res = torch.stack([img_w / 2., img_h / 2.], dim=-1)
        joints2d = joints2d / res[:, None, :]

        smpl_output['smpl_joints2d'] = joints2d
```
